src/monitor/poll_monitor.cpp

```diff
@@ -19,58 +19,80 @@
  * along with lo2s.  If not, see <http://www.gnu.org/licenses/>.
  */
 
-#include <lo2s/monitor/fd_monitor.hpp>
+#include <lo2s/config.hpp>
+#include <lo2s/error.hpp>
+#include <lo2s/monitor/poll_monitor.hpp>
 
-#include <nitro/lang/enumerate.hpp>
+extern "C"
+{
+#include <sys/timerfd.h>
+}
 
 namespace lo2s
 {
 namespace monitor
 {
-FdMonitor::FdMonitor(trace::Trace& trace, const std::string& name) : ThreadedMonitor(trace, name)
+PollMonitor::PollMonitor(trace::Trace& trace, const std::string& name)
+: ThreadedMonitor(trace, name)
 {
-    pfds_.resize(1);
+    pfds_.resize(2);
     stop_pfd().fd = stop_pipe_.read_fd();
     stop_pfd().events = POLLIN;
     stop_pfd().revents = 0;
+
+    // Create and initialize timer_fd
+    struct itimerspec tspec;
+    memset(&tspec, 0, sizeof(struct itimerspec));
+
+    // Set initial expiration to lowest possible value, this together with TFD_TIMER_ABSTIME should
+    // synchronize our timers
+    tspec.it_value.tv_nsec = 1;
+    tspec.it_interval.tv_nsec = config().read_interval.count();
+
+    timer_pfd().fd = timerfd_create(CLOCK_MONOTONIC, 0);
+    timer_pfd().events = POLLIN;
+    timer_pfd().revents = 0;
+
+    timerfd_settime(timer_pfd().fd, TFD_TIMER_ABSTIME, &tspec, NULL);
 }
 
-size_t FdMonitor::add_fd(int fd)
+void PollMonitor::add_fd(int fd)
 {
     struct pollfd pfd;
     pfd.fd = fd;
     pfd.events = POLLIN;
     pfd.revents = 0;
     pfds_.push_back(pfd);
-    return pfds_.size() - 2;
 }
 
-void FdMonitor::stop()
+void PollMonitor::stop()
 {
-    stop_pipe_.write();
     if (!thread_.joinable())
     {
-        Log::error() << "Cannot stop/join FdMonitor thread not running.";
+        return;
```

Comment:
This is carried over behaviour from the Intervalmonitor, where we kept zombie ThreadMonitors around in the ProcessMonitor to extract the cached mmap events. To avoid having to differentiate between zombie Monitors and alive ones when bulk stopping all ThreadMonitors at the end of execution, I have made stop idempotent (Which it should be anyways imho). I might wanna look into making this more sensible though, not declaring every thread that can't be joined as already stopped by us.

Reply:
Unless there currently is *known* good reason to call `stop` twice. I would emit a warning.

Reply:
As per today's discussion. Make `stop` virtual and implement the repeated stop / flag check only in the `ThreadMonitor`.